RecoTracker/TkTrackingRegions/plugins/CandidatePointSeededTrackingRegionsProducer.h

```diff
@@ -0,0 +1,451 @@
+
+    // this is a default origin for all modes
+    GlobalPoint default_origin( bs->x0(), bs->y0(), bs->z0() );
+
+    // vector of origin & halfLength pairs:
+    std::vector< std::pair< GlobalPoint, float > > origins;
+
+    // fill the origins and halfLengths depending on the mode
+    if (m_mode == BEAM_SPOT_FIXED || m_mode == BEAM_SPOT_SIGMA)
+    {
+      origins.push_back( std::make_pair(
+          default_origin,
+          (m_mode == BEAM_SPOT_FIXED) ? m_zErrorBeamSpot : m_nSigmaZBeamSpot*bs->sigmaZ()
+      ));
+    }
+    else if (m_mode == VERTICES_FIXED || m_mode == VERTICES_SIGMA)
+    {
+      edm::Handle< reco::VertexCollection > vertices;
+      e.getByToken( token_vertex, vertices );
+      int n_vert = 0;
+      for (reco::VertexCollection::const_iterator v = vertices->begin(); v != vertices->end() && n_vert < m_maxNVertices; ++v)
+      {
+        if ( v->isFake() || !v->isValid() ) continue;
+
+        origins.push_back( std::make_pair(
+            GlobalPoint( v->x(), v->y(), v->z() ),
+            (m_mode == VERTICES_FIXED) ? m_zErrorVetex : m_nSigmaZVertex*v->zError()
+        ));
+        ++n_vert;
+      }
+      // no-vertex fall-back case:
+      if (origins.empty())
+      {
+        origins.push_back( std::make_pair(
+            default_origin,
+            (m_nSigmaZBeamSpot > 0.) ? m_nSigmaZBeamSpot*bs->z0Error() : m_zErrorBeamSpot
+        ));
+      }
+    }
+    
+    const MeasurementTrackerEvent *measurementTracker = nullptr;
+    if(!token_measurementTracker.isUninitialized()) {
+      edm::Handle<MeasurementTrackerEvent> hmte;
+      e.getByToken(token_measurementTracker, hmte);
+      measurementTracker = hmte.product();
+    }
+
+    // create tracking regions (maximum MaxNRegions of them) in directions of the
+    // objects of interest (we expect that the collection was sorted in decreasing pt order)
+    int n_regions = 0;
+
+    if(m_seedingMode == CANDIDATE_SEEDED) {
+
+      for(size_t i = 0; i < n_objects && n_regions < m_maxNRegions; ++i ) {
+
+	const reco::Candidate & object = (*objects)[i];
+	GlobalVector direction( object.momentum().x(), object.momentum().y(), object.momentum().z() );
+	
+	for (size_t  j=0; j<origins.size() && n_regions < m_maxNRegions; ++j) {	
+
+	  result.push_back(std::make_unique<RectangularEtaPhiTrackingRegion>(
+		               direction,
+			       origins[j].first,
+			       m_ptMin,
+			       m_originRadius,
+			       origins[j].second,
+			       m_deltaEta,
+			       m_deltaPhi,
+			       m_whereToUseMeasurementTracker,
+			       m_precise,
+			       measurementTracker,
+			       m_searchOpt
+			   ));
+	  ++n_regions;
+
+	}
+
+      }
+
+    }
+     
+
+    else if(m_seedingMode == POINT_SEEDED) {
+
+      for(size_t i = 0; i < n_points && n_regions < m_maxNRegions; ++i ) {
+
+	double x = std::cos(phiPoints[i]);
+	double y = std::sin(phiPoints[i]);
+	double theta = 2*std::atan(std::exp(-etaPoints[i]));
+	double z = 1./std::tan(theta);
```

Comment:
These repeat the same calculations in all events, so one could calculate `x,y,z` once (well, actually the `GlobalVector direction` directly) in the constructor. On the other hand, for the `CANDIDATE_POINT_SEEDED` mode the phi and eta are still needed (and better be cached than calculated from `x,y,z` every time).